File: java/94.binary-tree-inorder-traversal.java
/*
 * @lc app=leetcode id=94 lang=java
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        in_order(root, result);
        return result;
    }

    public void in_order(TreeNode root, List<Integer> result){
        if(root == null) return;
        in_order(root.left, result);
        result.add(root.val);
        in_order(root.right, result);
    }
}
// @lc code=end

File: cpp/94.binary-tree-inorder-traversal.cpp
/*
 * @lc app=leetcode id=94 lang=cpp
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };

#include <vector>
using namespace std;
class Solution {
   public:
    vector<int> inorderTraversal(TreeNode *root) {
        // Note: Inorder traversal is left -> root -> right
        // i.e.: from smallest to largest
        vector<int> result;
        inorder_helper(root, result);
        return result;
    }

    void inorder_helper(TreeNode *root, vector<int> &result) {
        if (root == nullptr) {
            return;
        }
        // left first
        inorder_helper(root->left, result);
        // root
        result.push_back(root->val);
        // last right
        inorder_helper(root->right, result);
    }
};
// @lc code=end
